File: backend/GeoQuiz_backend/GeoQuiz_backend/Application/Payloads/Questions/CountryGrouping.cs
using GeoQuiz_backend.Domain.Mongo;

namespace GeoQuiz_backend.Application.Payloads.Questions
{
    public class CountryGrouping
    {
        public List<Country> AllCountries { get; set; }
        public ILookup<string, Country> CountriesByRegion { get; set; }
        public List<string> Regions { get; set; }

        public CountryGrouping(List<Country> allCountries)
        {
            AllCountries = allCountries;
            CountriesByRegion = allCountries.ToLookup(c => c.Region);
            Regions = CountriesByRegion.Select(g => g.Key).ToList();
        }
    }
}

File: frontend/app/src/main/java/com/example/geoquiz_frontend/domain/entities/CountryGrouping.java
package com.example.geoquiz_frontend.domain.entities;

import com.example.geoquiz_frontend.data.remote.dtos.solo.BootstrapResponse;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class CountryGrouping {
    private final List<BootstrapResponse.CountryDto> allCountries;
    private final Map<String, List<BootstrapResponse.CountryDto>> countriesByRegion;
    private final List<String> regions;

    public CountryGrouping(List<BootstrapResponse.CountryDto> allCountries) {
        this.allCountries = new ArrayList<>(allCountries);
        this.countriesByRegion = new HashMap<>();

        for (BootstrapResponse.CountryDto country : allCountries) {
            String region = country.getRegion();
            if (!countriesByRegion.containsKey(region)) {
                countriesByRegion.put(region, new ArrayList<>());
            }
            countriesByRegion.get(region).add(country);
        }

        this.regions = new ArrayList<>(countriesByRegion.keySet());
    }

    public List<BootstrapResponse.CountryDto> getAllCountries() {
        return allCountries;
    }

    public List<BootstrapResponse.CountryDto> getCountriesByRegion(String region) {
        return countriesByRegion.getOrDefault(region, new ArrayList<>());
    }

    public List<String> getRegions() {
        return regions;
    }
}
